www/app/blog/2024-11-19-distributed-ddl.mdx

```diff
@@ -0,0 +1,290 @@
+
+In order to resolve the state of transactions, we reconcile information from several sources:
+- `pg_prepared_xact` - This is where Postgres stores the state of prepared transactions. Every transaction in this table was successfully prepared but not yet committed. 
+- `distributed_transactions` - This is a table that Nile maintains on each database. We insert and update it as part of each DDL transaction. Rows here are only visible by other sessions after the transaction committed successfully, as part of Postgres' ACID guarantees which only allow reading committed rows.
+- `dist_tx_participants` - This table is in separate database. Nile uses this table to maintain overall state from all databases. It isn't guaranteed to always have the latest state. Therefore we use this table as an optimization, but we also check individual databases for newer information.
+- Transaction Coordinator in-memory state - The transaction coordinator maintains an in-memory cache with all the connections to the databases, the transaction IDs of active transactions and the status of these transactions. Unlike the other sources of information, this information is not persisted - if the coordinator crashes, we will recover this state from the other sources.
+
+Note that we use the existing implementation of transactions and ACID in Postgres to make sure we have all the state information that we need, persisted in the database. Postgres provides us with the basic primitives of atomicity, isolation and durability that makes the architecture reliable.
+ 
+During failure recovery, we collect the state from all these sources and reconcile it into a consistent state using few simple rules. Lets look at how these rules play out in two scenarios - a failed database and a failed transaction coordinator. 
+
+### Recovering from database failure
+
+This is a case where the transaction coordinator lost contact with a database. The first order of business is to get back in touch. Unfortunately, Postgres does not have "database is back up" hook that we can use for a callback. Therefore both the coordinator and the `nile_ddl` extension periodically check the connection and attempt to reconnect if there is an issue. Once a connection is established, the transaction coordinator will check `pg_prepared_xact` and `distributed_transactions` tables on the recovering database. For every open transaction in the coordinator's in-memory state, the recovering DB can be in one of 3 states:
+
+- Transaction committed successfuly. In this case, the coordinator just needs to update its in-memory state and the metadata state table. 
```

Comment:
Maybe I missed it but 'metadata state table' wasn't introduced above?

Reply:
Good catch. I didn't use this term.